File: sdks/swift/Sources/EdgeQuakeSDK/Services.swift
import Foundation

// MARK: - Service classes

/// WHY: Each service maps 1:1 to an API resource for discoverability.

public final class HealthService: @unchecked Sendable {
    private let http: HttpHelper
    init(_ http: HttpHelper) { self.http = http }

    public func check() async throws -> HealthResponse {
        try await http.get("/health")
    }
}

public final class DocumentService: @unchecked Sendable {
    private let http: HttpHelper
    init(_ http: HttpHelper) { self.http = http }

    public func list(page: Int = 1, pageSize: Int = 20) async throws -> ListDocumentsResponse {
        try await http.get("/api/v1/documents?page=\(page)&page_size=\(pageSize)")
    }

    public func get(id: String) async throws -> Document {
        try await http.get("/api/v1/documents/\(id)")
    }

    public func uploadText(title: String, content: String) async throws -> UploadResponse {
        try await http.post(
            "/api/v1/documents", body: TextUploadRequest(title: title, content: content))
    }

    /// WHY: DELETE returns 204 No Content — use deleteRaw to avoid decoding empty body.
    public func delete(id: String) async throws {
        _ = try await http.deleteRaw("/api/v1/documents/\(id)")
    }
}

public final class EntityService: @unchecked Sendable {
    private let http: HttpHelper
    init(_ http: HttpHelper) { self.http = http }

    public func list(page: Int = 1, pageSize: Int = 20) async throws -> EntityListResponse {
        try await http.get("/api/v1/graph/entities?page=\(page)&page_size=\(pageSize)")
    }

    public func get(name: String) async throws -> EntityDetailResponse {
        try await http.get("/api/v1/graph/entities/\(name)")
    }

    public func create(_ request: CreateEntityRequest) async throws -> CreateEntityResponse {
        try await http.post("/api/v1/graph/entities", body: request)
    }

    public func delete(name: String) async throws -> EntityDeleteResponse {
        try await http.delete("/api/v1/graph/entities/\(name)?confirm=true")
    }

    public func exists(name: String) async throws -> EntityExistsResponse {
        try await http.get("/api/v1/graph/entities/exists?entity_name=\(name)")
    }
}

public final class RelationshipService: @unchecked Sendable {
    private let http: HttpHelper
    init(_ http: HttpHelper) { self.http = http }

    public func list(page: Int = 1, pageSize: Int = 20) async throws -> RelationshipListResponse {
        try await http.get("/api/v1/graph/relationships?page=\(page)&page_size=\(pageSize)")
    }
}

public final class GraphService: @unchecked Sendable {
    private let http: HttpHelper
    init(_ http: HttpHelper) { self.http = http }

    public func get() async throws -> GraphResponse {
        try await http.get("/api/v1/graph")
    }

    public func search(query: String) async throws -> SearchNodesResponse {
        let encoded = query.addingPercentEncoding(withAllowedCharacters: .urlQueryAllowed) ?? query
        return try await http.get("/api/v1/graph/nodes/search?q=\(encoded)")
    }
}

public final class QueryService: @unchecked Sendable {
    private let http: HttpHelper
    init(_ http: HttpHelper) { self.http = http }

    public func execute(query: String, mode: String = "hybrid") async throws -> QueryResponse {
        try await http.post("/api/v1/query", body: QueryRequest(query: query, mode: mode))
    }
}

public final class ChatService: @unchecked Sendable {
    private let http: HttpHelper
    init(_ http: HttpHelper) { self.http = http }

    public func completions(_ request: ChatCompletionRequest) async throws -> ChatCompletionResponse
    {
        try await http.post("/api/v1/chat/completions", body: request)
    }
}

public final class TenantService: @unchecked Sendable {
    private let http: HttpHelper
    init(_ http: HttpHelper) { self.http = http }

    public func list() async throws -> TenantListResponse {
        try await http.get("/api/v1/tenants")
    }
}

public final class UserService: @unchecked Sendable {
    private let http: HttpHelper
    init(_ http: HttpHelper) { self.http = http }

    public func list() async throws -> UserListResponse {
        try await http.get("/api/v1/users")
    }
}

public final class ApiKeyService: @unchecked Sendable {
    private let http: HttpHelper
    init(_ http: HttpHelper) { self.http = http }

    public func list() async throws -> ApiKeyListResponse {
        try await http.get("/api/v1/api-keys")
    }
}

public final class TaskService: @unchecked Sendable {
    private let http: HttpHelper
    init(_ http: HttpHelper) { self.http = http }

    public func list() async throws -> TaskListResponse {
        try await http.get("/api/v1/tasks")
    }
}

public final class PipelineService: @unchecked Sendable {
    private let http: HttpHelper
    init(_ http: HttpHelper) { self.http = http }

    public func status() async throws -> PipelineStatus {
        try await http.get("/api/v1/pipeline/status")
    }

    public func queueMetrics() async throws -> QueueMetrics {
        try await http.get("/api/v1/pipeline/queue-metrics")
    }
}

public final class ModelService: @unchecked Sendable {
    private let http: HttpHelper
    init(_ http: HttpHelper) { self.http = http }

    public func catalog() async throws -> ProviderCatalog {
        try await http.get("/api/v1/models")
    }

    public func health() async throws -> [ProviderHealthInfo] {
        let data = try await http.getRaw("/api/v1/models/health")
        return try http.decodeJSON([ProviderHealthInfo].self, from: data)
    }

    public func providerStatus() async throws -> ProviderStatus {
        try await http.get("/api/v1/settings/provider/status")
    }
}

public final class CostService: @unchecked Sendable {
    private let http: HttpHelper
    init(_ http: HttpHelper) { self.http = http }

    public func summary() async throws -> CostSummary {
        try await http.get("/api/v1/costs/summary")
    }
}

public final class ConversationService: @unchecked Sendable {
    private let http: HttpHelper
    init(_ http: HttpHelper) { self.http = http }

    /// WHY: GET /api/v1/conversations returns {"items":[...]} wrapper, not raw array.
    public func list() async throws -> [ConversationInfo] {
        let wrapper: ConversationListResponse = try await http.get("/api/v1/conversations")
        return wrapper.items ?? []
    }

    public func create(title: String) async throws -> ConversationInfo {
        try await http.post("/api/v1/conversations", body: CreateConversationRequest(title: title))
    }

    public func get(id: String) async throws -> ConversationDetail {
        try await http.get("/api/v1/conversations/\(id)")
    }

    /// WHY: DELETE returns 204 No Content — use deleteRaw to avoid decoding empty body.
    public func delete(id: String) async throws {
        _ = try await http.deleteRaw("/api/v1/conversations/\(id)")
    }

    public func bulkDelete(ids: [String]) async throws -> BulkDeleteResponse {
        try await http.post("/api/v1/conversations/bulk/delete", body: ["ids": ids])
    }
}

public final class FolderService: @unchecked Sendable {
    private let http: HttpHelper
    init(_ http: HttpHelper) { self.http = http }

    public func list() async throws -> [FolderInfo] {
        try await http.get("/api/v1/folders")
    }

    public func create(name: String) async throws -> FolderInfo {
        try await http.post("/api/v1/folders", body: CreateFolderRequest(name: name))
    }

    /// WHY: DELETE returns 204 No Content — use deleteRaw to avoid decoding empty body.
    public func delete(id: String) async throws {
        _ = try await http.deleteRaw("/api/v1/folders/\(id)")
    }
}

File: sdks/kotlin/src/main/kotlin/io/edgequake/sdk/resources/Services.kt
package io.edgequake.sdk.resources

import com.fasterxml.jackson.core.type.TypeReference
import io.edgequake.sdk.internal.HttpHelper
import io.edgequake.sdk.models.*

/** WHY: Each service maps 1:1 to an API resource for discoverability. */

class HealthService(private val http: HttpHelper) {
    fun check(): HealthResponse = http.get("/health")
}

class DocumentService(private val http: HttpHelper) {
    fun list(page: Int = 1, pageSize: Int = 20): ListDocumentsResponse =
        http.get("/api/v1/documents?page=$page&page_size=$pageSize")

    fun get(id: String): Document = http.get("/api/v1/documents/$id")

    fun uploadText(title: String, content: String): UploadResponse {
        val json = http.postRaw("/api/v1/documents", TextUploadRequest(title, content))
        return http.mapper.readValue(json, UploadResponse::class.java)
    }

    /** WHY: DELETE may return 204 No Content — use deleteRaw to avoid deserialization of empty body. */
    fun delete(id: String) { http.deleteRaw("/api/v1/documents/$id") }

    fun scan(path: String, recursive: Boolean = true): ScanResponse =
        http.post("/api/v1/documents/scan", ScanRequest(path, recursive))
}

class EntityService(private val http: HttpHelper) {
    fun list(page: Int = 1, pageSize: Int = 20): EntityListResponse =
        http.get("/api/v1/graph/entities?page=$page&page_size=$pageSize")

    fun get(name: String): EntityDetailResponse =
        http.get("/api/v1/graph/entities/$name")

    fun create(req: CreateEntityRequest): CreateEntityResponse =
        http.post("/api/v1/graph/entities", req)

    fun delete(name: String): EntityDeleteResponse =
        http.delete("/api/v1/graph/entities/$name?confirm=true")

    fun exists(name: String): EntityExistsResponse =
        http.get("/api/v1/graph/entities/exists?entity_name=$name")

    fun merge(source: String, target: String): Map<String, Any?> =
        http.post("/api/v1/graph/entities/merge", MergeEntitiesRequest(source, target))
}

class RelationshipService(private val http: HttpHelper) {
    fun list(page: Int = 1, pageSize: Int = 20): RelationshipListResponse =
        http.get("/api/v1/graph/relationships?page=$page&page_size=$pageSize")
}

class GraphService(private val http: HttpHelper) {
    fun get(): GraphResponse = http.get("/api/v1/graph")

    fun search(query: String): SearchNodesResponse =
        http.get("/api/v1/graph/nodes/search?q=$query")
}

class QueryService(private val http: HttpHelper) {
    fun execute(query: String, mode: String = "hybrid"): QueryResponse =
        http.post("/api/v1/query", QueryRequest(query, mode))
}

class ChatService(private val http: HttpHelper) {
    fun completions(req: ChatCompletionRequest): ChatCompletionResponse =
        http.post("/api/v1/chat/completions", req)
}

class AuthService(private val http: HttpHelper) {
    fun login(username: String, password: String): TokenResponse =
        http.post("/api/v1/auth/login", LoginRequest(username, password))
}

class UserService(private val http: HttpHelper) {
    fun list(): UserListResponse = http.get("/api/v1/users")
}

class ApiKeyService(private val http: HttpHelper) {
    fun list(): ApiKeyListResponse = http.get("/api/v1/api-keys")
}

class TenantService(private val http: HttpHelper) {
    fun list(): TenantListResponse = http.get("/api/v1/tenants")
}

class ConversationService(private val http: HttpHelper) {
    /** WHY: GET /api/v1/conversations returns {"items":[...]} wrapper, not raw array. */
    fun list(): List<ConversationInfo> {
        val wrapper: ConversationListResponse = http.get("/api/v1/conversations")
        return wrapper.items ?: emptyList()
    }

    fun create(title: String): ConversationInfo =
        http.post("/api/v1/conversations", mapOf("title" to title))

    fun get(id: String): ConversationDetail = http.get("/api/v1/conversations/$id")

    /** WHY: DELETE returns 204 No Content — use deleteRaw to avoid deserialization of empty body. */
    fun delete(id: String) { http.deleteRaw("/api/v1/conversations/$id") }

    fun bulkDelete(ids: List<String>): BulkDeleteResponse =
        http.post("/api/v1/conversations/bulk/delete", mapOf("ids" to ids))
}

class FolderService(private val http: HttpHelper) {
    fun list(): List<FolderInfo> = http.get("/api/v1/folders")

    fun create(name: String): FolderInfo =
        http.post("/api/v1/folders", mapOf("name" to name))

    /** WHY: DELETE returns 204 No Content — use deleteRaw to avoid deserialization of empty body. */
    fun delete(id: String) { http.deleteRaw("/api/v1/folders/$id") }
}

class TaskService(private val http: HttpHelper) {
    fun list(): TaskListResponse = http.get("/api/v1/tasks")

    fun get(id: String): TaskInfo = http.get("/api/v1/tasks/$id")
}

class PipelineService(private val http: HttpHelper) {
    fun status(): PipelineStatus = http.get("/api/v1/pipeline/status")

    fun queueMetrics(): QueueMetrics = http.get("/api/v1/pipeline/queue-metrics")
}

class ModelService(private val http: HttpHelper) {
    fun catalog(): ProviderCatalog = http.get("/api/v1/models")

    fun health(): List<ProviderHealthInfo> {
        val json = http.getRaw("/api/v1/models/health")
        return http.mapper.readValue(json, object : TypeReference<List<ProviderHealthInfo>>() {})
    }

    fun providerStatus(): ProviderStatus =
        http.get("/api/v1/settings/provider/status")
}

class WorkspaceService(private val http: HttpHelper) {
    fun list(): List<WorkspaceInfo> = http.get("/api/v1/workspaces")
}

class PdfService(private val http: HttpHelper) {
    fun progress(trackId: String): PdfProgressResponse =
        http.get("/api/v1/documents/pdf/progress/$trackId")

    fun content(pdfId: String): PdfContentResponse =
        http.get("/api/v1/documents/pdf/$pdfId/content")
}

class CostService(private val http: HttpHelper) {
    fun summary(): CostSummary = http.get("/api/v1/costs/summary")
}
